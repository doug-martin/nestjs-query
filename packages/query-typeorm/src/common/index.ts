export * from './randomString';
export * from './typeorm';
