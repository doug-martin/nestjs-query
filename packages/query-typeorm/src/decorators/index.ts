export { InjectTypeOrmQueryService } from './inject-typeorm-query-service.decorator';
