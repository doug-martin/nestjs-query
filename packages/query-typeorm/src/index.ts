export { TypeOrmQueryService, TypeOrmQueryServiceOpts } from './services';
export { NestjsQueryTypeOrmModule } from './module';
