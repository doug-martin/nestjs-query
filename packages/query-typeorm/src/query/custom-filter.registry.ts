import { Class } from '@nestjs-query/core';
import { ColumnType, ObjectLiteral } from 'typeorm';

export type CustomFilterResult = { sql: string; params: ObjectLiteral };

export interface CustomFilterContext {
  fieldType?: ColumnType;
}

export interface CustomFilter<OperationType = string, Entity = unknown> {
  apply(
    field: keyof Entity | string,
    cmp: OperationType,
    val: unknown,
    alias?: string,
    context?: CustomFilterContext,
  ): CustomFilterResult;
}

type OperationCustomFilters = Record<string, CustomFilter>;

type SetFilterSpec =
  | { types: ColumnType[]; operations: string[] }
  | { klass: Class<unknown>; field: string; operations: string[] };

export class CustomFilterRegistry {
  // Registry for (class, field) filters
  private cfoRegistry: Map<Class<unknown>, Record<string | symbol | number, OperationCustomFilters>> = new Map();

  // Registry for (type) filters
  private tRegistry: Map<ColumnType, OperationCustomFilters> = new Map();

  getFilter<Entity = unknown>(
    operation: string,
    type?: ColumnType, // Type is optional, since we might have non property-backed filters
    klass?: Class<Entity>,
    field?: keyof Entity | string,
  ): CustomFilter<string, Entity> | undefined {
    // Most specific: (class, field) filters.
    if (klass && field) {
      const flt = this.cfoRegistry.get(klass)?.[field]?.[operation];
      if (flt) {
        return flt;
      }
    }
    // Type filters
    if (type) {
      return this.tRegistry.get(type)?.[operation];
    }
    return undefined;
  }

  /**
   * We have 2 types of filters:
   * - (type, operation) global filters
   * - (class, field, operation) filters
   * type is the database column type (TypeORM's ColumnType)
   * Specificity of the filters increases from top to bottom
   */
  setFilter<Entity>(filter: CustomFilter<string, Entity>, opts: SetFilterSpec): void {
    if ('klass' in opts && 'field' in opts) {
      const { klass, field, operations } = opts;
      if (!operations || operations.length === 0) {
        throw new Error('Cannot register a filter without operations, please define the operations array');
      }
      const entityFilters = this.cfoRegistry.get(klass) || {};
      entityFilters[field] = this.createCustomFilterOperationMap(filter, operations);
      this.cfoRegistry.set(klass, entityFilters);
    } else if ('types' in opts) {
      const { types, operations } = opts;
      if (!types || types.length === 0) {
        throw new Error('Cannot register a (type) filter without types, please define the types array');
      }
      if (!operations || operations.length === 0) {
        throw new Error('Cannot register a filter without operations, please define the operations array');
      }
      for (const type of types) {
        this.tRegistry.set(type, this.createCustomFilterOperationMap(filter, operations));
      }
    }
  }

  private createCustomFilterOperationMap(cf: CustomFilter, operations: string[]): OperationCustomFilters {
    const ocf: OperationCustomFilters = {};
    for (const op of operations) {
      ocf[op] = cf;
    }
    return ocf;
  }
}
