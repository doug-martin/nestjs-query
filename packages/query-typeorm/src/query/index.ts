export * from './filter-query.builder';
export * from './where.builder';
export * from './sql-comparison.builder';
export * from './relation-query.builder';
export * from './aggregate.builder';
