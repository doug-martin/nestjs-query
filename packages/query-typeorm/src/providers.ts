import { Class, getQueryServiceToken } from '@nestjs-query/core';
import { FactoryProvider } from '@nestjs/common';
import { getRepositoryToken } from '@nestjs/typeorm';
import { Repository, Connection, ConnectionOptions } from 'typeorm';
import { TypeOrmQueryService } from './services';

function createTypeOrmQueryServiceProvider<Entity>(
  EntityClass: Class<Entity>,
  connection?: Connection | ConnectionOptions | string,
): FactoryProvider {
  return {
    provide: getQueryServiceToken(EntityClass),
    useFactory(repo: Repository<Entity>) {
      return new TypeOrmQueryService(repo);
    },
    inject: [getRepositoryToken(EntityClass, connection)],
  };
}

export const createTypeOrmQueryServiceProviders = (
  entities: Class<unknown>[],
  connection?: Connection | ConnectionOptions | string,
): FactoryProvider[] => entities.map((entity) => createTypeOrmQueryServiceProvider(entity, connection));
