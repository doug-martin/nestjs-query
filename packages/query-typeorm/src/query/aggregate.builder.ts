import { SelectQueryBuilder } from 'typeorm';
import { AggregateQuery, AggregateResponse } from '@nestjs-query/core';
import { BadRequestException } from '@nestjs/common';
import { camelCase } from 'camel-case';

enum AggregateFuncs {
  AVG = 'AVG',
  SUM = 'SUM',
  COUNT = 'COUNT',
  MAX = 'MAX',
  MIN = 'MIN',
}

const AGG_REGEXP = /(AVG|SUM|COUNT|MAX|MIN|GROUP_BY)_(.*)/;

/**
 * @internal
 * Builds a WHERE clause from a Filter.
 */
export class AggregateBuilder<Entity> {
  // eslint-disable-next-line @typescript-eslint/no-shadow
  static async asyncConvertToAggregateResponse<Entity>(
    responsePromise: Promise<Record<string, unknown>[]>,
  ): Promise<AggregateResponse<Entity>[]> {
    const aggResponse = await responsePromise;
    return this.convertToAggregateResponse(aggResponse);
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  static getAggregateSelects<Entity>(query: AggregateQuery<Entity>): string[] {
    return [...this.getAggregateGroupBySelects(query), ...this.getAggregateFuncSelects(query)];
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  private static getAggregateGroupBySelects<Entity>(query: AggregateQuery<Entity>): string[] {
    return (query.groupBy ?? []).map((f) => this.getGroupByAlias(f));
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  private static getAggregateFuncSelects<Entity>(query: AggregateQuery<Entity>): string[] {
    const aggs: [AggregateFuncs, (keyof Entity)[] | undefined][] = [
      [AggregateFuncs.COUNT, query.count],
      [AggregateFuncs.SUM, query.sum],
      [AggregateFuncs.AVG, query.avg],
      [AggregateFuncs.MAX, query.max],
      [AggregateFuncs.MIN, query.min],
    ];
    return aggs.reduce((cols, [func, fields]) => {
      const aliases = (fields ?? []).map((f) => this.getAggregateAlias(func, f));
      return [...cols, ...aliases];
    }, [] as string[]);
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  static getAggregateAlias<Entity>(func: AggregateFuncs, field: keyof Entity): string {
    return `${func}_${field as string}`;
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  static getGroupByAlias<Entity>(field: keyof Entity): string {
    return `GROUP_BY_${field as string}`;
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  static convertToAggregateResponse<Entity>(rawAggregates: Record<string, unknown>[]): AggregateResponse<Entity>[] {
    return rawAggregates.map((response) => {
      return Object.keys(response).reduce((agg: AggregateResponse<Entity>, resultField: string) => {
        const matchResult = AGG_REGEXP.exec(resultField);
        if (!matchResult) {
          throw new Error('Unknown aggregate column encountered.');
        }
        const [matchedFunc, matchedFieldName] = matchResult.slice(1);
        const aggFunc = camelCase(matchedFunc.toLowerCase()) as keyof AggregateResponse<Entity>;
        const fieldName = matchedFieldName as keyof Entity;
        const aggResult = agg[aggFunc] || {};
        return {
          ...agg,
          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
          [aggFunc]: { ...aggResult, [fieldName]: response[resultField] },
        };
      }, {});
    });
  }

  /**
   * Builds a aggregate SELECT clause from a aggregate.
   * @param qb - the `typeorm` SelectQueryBuilder
   * @param aggregate - the aggregates to select.
   * @param alias - optional alias to use to qualify an identifier
   */
  build<Qb extends SelectQueryBuilder<Entity>>(qb: Qb, aggregate: AggregateQuery<Entity>, alias?: string): Qb {
    const selects = [
      ...this.createGroupBySelect(aggregate.groupBy, alias),
      ...this.createAggSelect(AggregateFuncs.COUNT, aggregate.count, alias),
      ...this.createAggSelect(AggregateFuncs.SUM, aggregate.sum, alias),
      ...this.createAggSelect(AggregateFuncs.AVG, aggregate.avg, alias),
      ...this.createAggSelect(AggregateFuncs.MAX, aggregate.max, alias),
      ...this.createAggSelect(AggregateFuncs.MIN, aggregate.min, alias),
    ];
    if (!selects.length) {
      throw new BadRequestException('No aggregate fields found.');
    }
    const [head, ...tail] = selects;
    return tail.reduce(
      (acc: Qb, [select, selectAlias]) => acc.addSelect(select, selectAlias),
      qb.select(head[0], head[1]),
    );
  }

  private createAggSelect(func: AggregateFuncs, fields?: (keyof Entity)[], alias?: string): [string, string][] {
    if (!fields) {
      return [];
    }
    return fields.map((field) => {
      const col = alias ? `${alias}.${field as string}` : (field as string);
      return [`${func}(${col})`, AggregateBuilder.getAggregateAlias(func, field)];
    });
  }

  private createGroupBySelect(fields?: (keyof Entity)[], alias?: string): [string, string][] {
    if (!fields) {
      return [];
    }
    return fields.map((field) => {
      const col = alias ? `${alias}.${field as string}` : (field as string);
      return [`${col}`, AggregateBuilder.getGroupByAlias(field)];
    });
  }
}
