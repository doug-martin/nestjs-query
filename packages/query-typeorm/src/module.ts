import { Class } from '@nestjs-query/core';
import { TypeOrmModule } from '@nestjs/typeorm';
import { DynamicModule } from '@nestjs/common';
import { Connection, ConnectionOptions } from 'typeorm';
import { createTypeOrmQueryServiceProviders } from './providers';

export class NestjsQueryTypeOrmModule {
  static forFeature(entities: Class<unknown>[], connection?: Connection | ConnectionOptions | string): DynamicModule {
    const queryServiceProviders = createTypeOrmQueryServiceProviders(entities, connection);
    const typeOrmModule = TypeOrmModule.forFeature(entities, connection);
    return {
      imports: [typeOrmModule],
      module: NestjsQueryTypeOrmModule,
      providers: [...queryServiceProviders],
      exports: [...queryServiceProviders, typeOrmModule],
    };
  }
}
