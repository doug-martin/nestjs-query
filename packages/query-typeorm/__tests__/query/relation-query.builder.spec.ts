import { Class, Query, SortDirection, SortNulls } from '@nestjs-query/core';
import { closeTestConnection, createTestConnection, getTestConnection } from '../__fixtures__/connection.fixture';
import { TestRelation } from '../__fixtures__/test-relation.entity';
import { TestEntity } from '../__fixtures__/test.entity';
import { RelationQueryBuilder } from '../../src/query';

describe('RelationQueryBuilder', (): void => {
  beforeEach(createTestConnection);
  afterEach(closeTestConnection);

  const getRelationQueryBuilder = <Entity, Relation>(
    EntityClass: Class<Entity>,
    relationName: string,
  ): RelationQueryBuilder<Entity, Relation> =>
    new RelationQueryBuilder(getTestConnection().getRepository(EntityClass), relationName);

  const expectSQLSnapshot = <Entity, Relation>(
    EntityClass: Class<Entity>,
    entity: Entity,
    relation: string,
    query: Query<Relation>,
  ): void => {
    const selectQueryBuilder = getRelationQueryBuilder<Entity, Relation>(EntityClass, relation).select(entity, query);
    const [sql, params] = selectQueryBuilder.getQueryAndParameters();
    expect(sql).toMatchSnapshot();
    expect(params).toMatchSnapshot();
  };

  describe('#select', () => {
    const testEntity: TestEntity = {
      testEntityPk: 'test-entity-id-1',
      dateType: new Date(),
      boolType: true,
      numberType: 1,
      stringType: 'str',
    };

    const testRelation: TestRelation = {
      testRelationPk: 'test-relation-id-1',
      relationName: 'relation-name',
    };

    it('should throw an error if there is no relation with that name', () => {
      expect(() => {
        expectSQLSnapshot(TestEntity, testEntity, 'badRelations', {});
      }).toThrow("Unable to find entity for relation 'badRelations'");
    });

    describe('one to many', () => {
      it('should query with a single entity', () => {
        expectSQLSnapshot(TestEntity, testEntity, 'testRelations', {});
      });
    });

    describe('many to one', () => {
      it('should work with one entity', () => {
        expectSQLSnapshot(TestRelation, testRelation, 'testEntity', {});
      });

      it('should work with a uni-directional relationship', () => {
        expectSQLSnapshot(TestRelation, testRelation, 'testEntityUniDirectional', {});
      });
    });

    describe('many to many', () => {
      describe('on owning side', () => {
        it('should work with one entity', () => {
          expectSQLSnapshot(TestEntity, testEntity, 'manyTestRelations', {});
        });
      });

      describe('on non owning side', () => {
        it('should work with many to many', () => {
          expectSQLSnapshot(TestRelation, testRelation, 'manyTestEntities', {});
        });
      });

      describe('many-to-many custom join table', () => {
        it('should work with a many-to-many through a join table', () => {
          expectSQLSnapshot(TestEntity, testEntity, 'testEntityRelation', {});
        });
      });

      describe('uni-directional many to many', () => {
        it('should create the correct sql', () => {
          expectSQLSnapshot(TestEntity, testEntity, 'manyToManyUniDirectional', {});
        });
      });
    });

    describe('one to one', () => {
      it('on owning side', () => {
        expectSQLSnapshot(TestEntity, testEntity, 'oneTestRelation', {});
      });

      it('on non owning side', () => {
        expectSQLSnapshot(TestRelation, testRelation, 'oneTestEntity', {});
      });
    });

    describe('with filter', () => {
      it('should call whereBuilder#build if there is a filter', () => {
        const query: Query<TestRelation> = { filter: { relationName: { eq: 'foo' } } };
        expectSQLSnapshot(TestEntity, testEntity, 'testRelations', query);
      });
    });

    describe('with paging', () => {
      it('should apply paging args going forward', () => {
        expectSQLSnapshot(TestEntity, testEntity, 'testRelations', { paging: { limit: 10, offset: 11 } });
      });

      it('should apply paging args going backward', () => {
        expectSQLSnapshot(TestEntity, testEntity, 'testRelations', { paging: { limit: 10, offset: 10 } });
      });
    });

    describe('with sorting', () => {
      it('should apply ASC sorting', () => {
        expectSQLSnapshot(TestEntity, testEntity, 'testRelations', {
          sorting: [{ field: 'relationName', direction: SortDirection.ASC }],
        });
      });

      it('should apply ASC NULLS_FIRST sorting', () => {
        expectSQLSnapshot(TestEntity, testEntity, 'testRelations', {
          sorting: [{ field: 'relationName', direction: SortDirection.ASC, nulls: SortNulls.NULLS_FIRST }],
        });
      });

      it('should apply ASC NULLS_LAST sorting', () => {
        expectSQLSnapshot(TestEntity, testEntity, 'testRelations', {
          sorting: [{ field: 'relationName', direction: SortDirection.ASC, nulls: SortNulls.NULLS_LAST }],
        });
      });

      it('should apply DESC sorting', () => {
        expectSQLSnapshot(TestEntity, testEntity, 'testRelations', {
          sorting: [{ field: 'relationName', direction: SortDirection.DESC }],
        });
      });

      it('should apply DESC NULLS_FIRST sorting', () => {
        expectSQLSnapshot(TestEntity, testEntity, 'testRelations', {
          sorting: [{ field: 'relationName', direction: SortDirection.DESC, nulls: SortNulls.NULLS_FIRST }],
        });
      });

      it('should apply DESC NULLS_LAST sorting', () => {
        expectSQLSnapshot(TestEntity, testEntity, 'testRelations', {
          sorting: [{ field: 'relationName', direction: SortDirection.DESC, nulls: SortNulls.NULLS_LAST }],
        });
      });

      it('should apply multiple sorts', () => {
        expectSQLSnapshot(TestEntity, testEntity, 'testRelations', {
          sorting: [
            { field: 'relationName', direction: SortDirection.ASC },
            { field: 'testRelationPk', direction: SortDirection.DESC },
          ],
        });
      });
    });
  });
});
