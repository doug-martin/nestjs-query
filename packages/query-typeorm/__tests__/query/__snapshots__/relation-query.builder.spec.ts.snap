// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelationQueryBuilder #select many to many many-to-many custom join table should work with a many-to-many through a join table 1`] = `SELECT "testEntityRelation"."test_relation_id" AS "testEntityRelation_test_relation_id", "testEntityRelation"."test_entity_id" AS "testEntityRelation_test_entity_id" FROM "test_entity_relation_entity" "testEntityRelation" WHERE ("testEntityRelation"."test_entity_id" = ?)`;

exports[`RelationQueryBuilder #select many to many many-to-many custom join table should work with a many-to-many through a join table 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select many to many on non owning side should work with many to many 1`] = `SELECT "manyTestEntities"."test_entity_pk" AS "manyTestEntities_test_entity_pk", "manyTestEntities"."string_type" AS "manyTestEntities_string_type", "manyTestEntities"."bool_type" AS "manyTestEntities_bool_type", "manyTestEntities"."number_type" AS "manyTestEntities_number_type", "manyTestEntities"."date_type" AS "manyTestEntities_date_type", "manyTestEntities"."oneTestRelationTestRelationPk" AS "manyTestEntities_oneTestRelationTestRelationPk" FROM "test_entity" "manyTestEntities" INNER JOIN "test_entity_many_test_relations_test_relation" "test_entity_many_test_relations_test_relation" ON "test_entity_many_test_relations_test_relation"."testEntityTestEntityPk" = "manyTestEntities"."test_entity_pk" WHERE ("test_entity_many_test_relations_test_relation"."testRelationTestRelationPk" = ?)`;

exports[`RelationQueryBuilder #select many to many on non owning side should work with many to many 2`] = `
Array [
  test-relation-id-1,
]
`;

exports[`RelationQueryBuilder #select many to many on owning side should work with one entity 1`] = `SELECT "manyTestRelations"."test_relation_pk" AS "manyTestRelations_test_relation_pk", "manyTestRelations"."relation_name" AS "manyTestRelations_relation_name", "manyTestRelations"."test_entity_id" AS "manyTestRelations_test_entity_id", "manyTestRelations"."uni_directional_test_entity_id" AS "manyTestRelations_uni_directional_test_entity_id", "manyTestRelations"."uni_directional_relation_test_entity_id" AS "manyTestRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "manyTestRelations" INNER JOIN "test_entity_many_test_relations_test_relation" "test_entity_many_test_relations_test_relation" ON "test_entity_many_test_relations_test_relation"."testRelationTestRelationPk" = "manyTestRelations"."test_relation_pk" WHERE ("test_entity_many_test_relations_test_relation"."testEntityTestEntityPk" = ?)`;

exports[`RelationQueryBuilder #select many to many on owning side should work with one entity 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select many to many uni-directional many to many should create the correct sql 1`] = `SELECT "manyToManyUniDirectional"."test_relation_pk" AS "manyToManyUniDirectional_test_relation_pk", "manyToManyUniDirectional"."relation_name" AS "manyToManyUniDirectional_relation_name", "manyToManyUniDirectional"."test_entity_id" AS "manyToManyUniDirectional_test_entity_id", "manyToManyUniDirectional"."uni_directional_test_entity_id" AS "manyToManyUniDirectional_uni_directional_test_entity_id", "manyToManyUniDirectional"."uni_directional_relation_test_entity_id" AS "manyToManyUniDirectional_uni_directional_relation_test_entity_id" FROM "test_relation" "manyToManyUniDirectional" INNER JOIN "test_entity_many_to_many_uni_directional_test_relation" "test_entity_many_to_many_uni_directional_test_relation" ON "test_entity_many_to_many_uni_directional_test_relation"."testRelationTestRelationPk" = "manyToManyUniDirectional"."test_relation_pk" WHERE ("test_entity_many_to_many_uni_directional_test_relation"."testEntityTestEntityPk" = ?)`;

exports[`RelationQueryBuilder #select many to many uni-directional many to many should create the correct sql 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select many to one should work with a uni-directional relationship 1`] = `SELECT "testEntityUniDirectional"."test_entity_pk" AS "testEntityUniDirectional_test_entity_pk", "testEntityUniDirectional"."string_type" AS "testEntityUniDirectional_string_type", "testEntityUniDirectional"."bool_type" AS "testEntityUniDirectional_bool_type", "testEntityUniDirectional"."number_type" AS "testEntityUniDirectional_number_type", "testEntityUniDirectional"."date_type" AS "testEntityUniDirectional_date_type", "testEntityUniDirectional"."oneTestRelationTestRelationPk" AS "testEntityUniDirectional_oneTestRelationTestRelationPk" FROM "test_entity" "testEntityUniDirectional" INNER JOIN "test_relation" "TestRelation" ON "TestRelation"."uni_directional_test_entity_id" = "testEntityUniDirectional"."test_entity_pk" WHERE ("TestRelation"."test_relation_pk" = ?)`;

exports[`RelationQueryBuilder #select many to one should work with a uni-directional relationship 2`] = `
Array [
  test-relation-id-1,
]
`;

exports[`RelationQueryBuilder #select many to one should work with one entity 1`] = `SELECT "testEntity"."test_entity_pk" AS "testEntity_test_entity_pk", "testEntity"."string_type" AS "testEntity_string_type", "testEntity"."bool_type" AS "testEntity_bool_type", "testEntity"."number_type" AS "testEntity_number_type", "testEntity"."date_type" AS "testEntity_date_type", "testEntity"."oneTestRelationTestRelationPk" AS "testEntity_oneTestRelationTestRelationPk" FROM "test_entity" "testEntity" INNER JOIN "test_relation" "TestRelation" ON "TestRelation"."test_entity_id" = "testEntity"."test_entity_pk" WHERE ("TestRelation"."test_relation_pk" = ?)`;

exports[`RelationQueryBuilder #select many to one should work with one entity 2`] = `
Array [
  test-relation-id-1,
]
`;

exports[`RelationQueryBuilder #select one to many should query with a single entity 1`] = `SELECT "testRelations"."test_relation_pk" AS "testRelations_test_relation_pk", "testRelations"."relation_name" AS "testRelations_relation_name", "testRelations"."test_entity_id" AS "testRelations_test_entity_id", "testRelations"."uni_directional_test_entity_id" AS "testRelations_uni_directional_test_entity_id", "testRelations"."uni_directional_relation_test_entity_id" AS "testRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "testRelations" WHERE ("testRelations"."test_entity_id" = ?)`;

exports[`RelationQueryBuilder #select one to many should query with a single entity 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select one to one on non owning side 1`] = `SELECT "oneTestEntity"."test_entity_pk" AS "oneTestEntity_test_entity_pk", "oneTestEntity"."string_type" AS "oneTestEntity_string_type", "oneTestEntity"."bool_type" AS "oneTestEntity_bool_type", "oneTestEntity"."number_type" AS "oneTestEntity_number_type", "oneTestEntity"."date_type" AS "oneTestEntity_date_type", "oneTestEntity"."oneTestRelationTestRelationPk" AS "oneTestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "oneTestEntity" WHERE ("oneTestEntity"."oneTestRelationTestRelationPk" = ?)`;

exports[`RelationQueryBuilder #select one to one on non owning side 2`] = `
Array [
  test-relation-id-1,
]
`;

exports[`RelationQueryBuilder #select one to one on owning side 1`] = `SELECT "oneTestRelation"."test_relation_pk" AS "oneTestRelation_test_relation_pk", "oneTestRelation"."relation_name" AS "oneTestRelation_relation_name", "oneTestRelation"."test_entity_id" AS "oneTestRelation_test_entity_id", "oneTestRelation"."uni_directional_test_entity_id" AS "oneTestRelation_uni_directional_test_entity_id", "oneTestRelation"."uni_directional_relation_test_entity_id" AS "oneTestRelation_uni_directional_relation_test_entity_id" FROM "test_relation" "oneTestRelation" INNER JOIN "test_entity" "TestEntity" ON "TestEntity"."oneTestRelationTestRelationPk" = "oneTestRelation"."test_relation_pk" WHERE ("TestEntity"."test_entity_pk" = ?)`;

exports[`RelationQueryBuilder #select one to one on owning side 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select with filter should call whereBuilder#build if there is a filter 1`] = `SELECT "testRelations"."test_relation_pk" AS "testRelations_test_relation_pk", "testRelations"."relation_name" AS "testRelations_relation_name", "testRelations"."test_entity_id" AS "testRelations_test_entity_id", "testRelations"."uni_directional_test_entity_id" AS "testRelations_uni_directional_test_entity_id", "testRelations"."uni_directional_relation_test_entity_id" AS "testRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "testRelations" WHERE ("testRelations"."test_entity_id" = ?) AND ("testRelations"."relation_name" = ?)`;

exports[`RelationQueryBuilder #select with filter should call whereBuilder#build if there is a filter 2`] = `
Array [
  test-entity-id-1,
  foo,
]
`;

exports[`RelationQueryBuilder #select with paging should apply paging args going backward 1`] = `SELECT "testRelations"."test_relation_pk" AS "testRelations_test_relation_pk", "testRelations"."relation_name" AS "testRelations_relation_name", "testRelations"."test_entity_id" AS "testRelations_test_entity_id", "testRelations"."uni_directional_test_entity_id" AS "testRelations_uni_directional_test_entity_id", "testRelations"."uni_directional_relation_test_entity_id" AS "testRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "testRelations" WHERE ("testRelations"."test_entity_id" = ?) LIMIT 10 OFFSET 10`;

exports[`RelationQueryBuilder #select with paging should apply paging args going backward 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select with paging should apply paging args going forward 1`] = `SELECT "testRelations"."test_relation_pk" AS "testRelations_test_relation_pk", "testRelations"."relation_name" AS "testRelations_relation_name", "testRelations"."test_entity_id" AS "testRelations_test_entity_id", "testRelations"."uni_directional_test_entity_id" AS "testRelations_uni_directional_test_entity_id", "testRelations"."uni_directional_relation_test_entity_id" AS "testRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "testRelations" WHERE ("testRelations"."test_entity_id" = ?) LIMIT 10 OFFSET 11`;

exports[`RelationQueryBuilder #select with paging should apply paging args going forward 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select with sorting should apply ASC NULLS_FIRST sorting 1`] = `SELECT "testRelations"."test_relation_pk" AS "testRelations_test_relation_pk", "testRelations"."relation_name" AS "testRelations_relation_name", "testRelations"."test_entity_id" AS "testRelations_test_entity_id", "testRelations"."uni_directional_test_entity_id" AS "testRelations_uni_directional_test_entity_id", "testRelations"."uni_directional_relation_test_entity_id" AS "testRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "testRelations" WHERE ("testRelations"."test_entity_id" = ?) ORDER BY "testRelations"."relation_name" ASC NULLS FIRST`;

exports[`RelationQueryBuilder #select with sorting should apply ASC NULLS_FIRST sorting 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select with sorting should apply ASC NULLS_LAST sorting 1`] = `SELECT "testRelations"."test_relation_pk" AS "testRelations_test_relation_pk", "testRelations"."relation_name" AS "testRelations_relation_name", "testRelations"."test_entity_id" AS "testRelations_test_entity_id", "testRelations"."uni_directional_test_entity_id" AS "testRelations_uni_directional_test_entity_id", "testRelations"."uni_directional_relation_test_entity_id" AS "testRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "testRelations" WHERE ("testRelations"."test_entity_id" = ?) ORDER BY "testRelations"."relation_name" ASC NULLS LAST`;

exports[`RelationQueryBuilder #select with sorting should apply ASC NULLS_LAST sorting 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select with sorting should apply ASC sorting 1`] = `SELECT "testRelations"."test_relation_pk" AS "testRelations_test_relation_pk", "testRelations"."relation_name" AS "testRelations_relation_name", "testRelations"."test_entity_id" AS "testRelations_test_entity_id", "testRelations"."uni_directional_test_entity_id" AS "testRelations_uni_directional_test_entity_id", "testRelations"."uni_directional_relation_test_entity_id" AS "testRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "testRelations" WHERE ("testRelations"."test_entity_id" = ?) ORDER BY "testRelations"."relation_name" ASC`;

exports[`RelationQueryBuilder #select with sorting should apply ASC sorting 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select with sorting should apply DESC NULLS_FIRST sorting 1`] = `SELECT "testRelations"."test_relation_pk" AS "testRelations_test_relation_pk", "testRelations"."relation_name" AS "testRelations_relation_name", "testRelations"."test_entity_id" AS "testRelations_test_entity_id", "testRelations"."uni_directional_test_entity_id" AS "testRelations_uni_directional_test_entity_id", "testRelations"."uni_directional_relation_test_entity_id" AS "testRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "testRelations" WHERE ("testRelations"."test_entity_id" = ?) ORDER BY "testRelations"."relation_name" DESC NULLS FIRST`;

exports[`RelationQueryBuilder #select with sorting should apply DESC NULLS_FIRST sorting 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select with sorting should apply DESC NULLS_LAST sorting 1`] = `SELECT "testRelations"."test_relation_pk" AS "testRelations_test_relation_pk", "testRelations"."relation_name" AS "testRelations_relation_name", "testRelations"."test_entity_id" AS "testRelations_test_entity_id", "testRelations"."uni_directional_test_entity_id" AS "testRelations_uni_directional_test_entity_id", "testRelations"."uni_directional_relation_test_entity_id" AS "testRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "testRelations" WHERE ("testRelations"."test_entity_id" = ?) ORDER BY "testRelations"."relation_name" DESC NULLS LAST`;

exports[`RelationQueryBuilder #select with sorting should apply DESC NULLS_LAST sorting 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select with sorting should apply DESC sorting 1`] = `SELECT "testRelations"."test_relation_pk" AS "testRelations_test_relation_pk", "testRelations"."relation_name" AS "testRelations_relation_name", "testRelations"."test_entity_id" AS "testRelations_test_entity_id", "testRelations"."uni_directional_test_entity_id" AS "testRelations_uni_directional_test_entity_id", "testRelations"."uni_directional_relation_test_entity_id" AS "testRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "testRelations" WHERE ("testRelations"."test_entity_id" = ?) ORDER BY "testRelations"."relation_name" DESC`;

exports[`RelationQueryBuilder #select with sorting should apply DESC sorting 2`] = `
Array [
  test-entity-id-1,
]
`;

exports[`RelationQueryBuilder #select with sorting should apply multiple sorts 1`] = `SELECT "testRelations"."test_relation_pk" AS "testRelations_test_relation_pk", "testRelations"."relation_name" AS "testRelations_relation_name", "testRelations"."test_entity_id" AS "testRelations_test_entity_id", "testRelations"."uni_directional_test_entity_id" AS "testRelations_uni_directional_test_entity_id", "testRelations"."uni_directional_relation_test_entity_id" AS "testRelations_uni_directional_relation_test_entity_id" FROM "test_relation" "testRelations" WHERE ("testRelations"."test_entity_id" = ?) ORDER BY "testRelations"."relation_name" ASC, "testRelations"."test_relation_pk" DESC`;

exports[`RelationQueryBuilder #select with sorting should apply multiple sorts 2`] = `
Array [
  test-entity-id-1,
]
`;
